### PR TITLE
Add Prometheus metrics validation

### PR DESCRIPTION
**Pull request by doniacld**
## Motivation
Add the possibility to verify metrics from the Cilium agent pod and validate it has the expected behaviour before and after an action. 

## Changes

- [commit 1](https://github.com/cilium/cilium-cli/pull/1575/commits/36277d8e22cc18328b8422eb2ac91939aa89e451) Add a k8s dialer to be able to port forward `/metrics` endpoint outside of the pod
- [commit 2](https://github.com/cilium/cilium-cli/pull/1575/commits/8d5197c451fd056e345337a3d775ee059f2d8a75) Add the logic to call `/metrics` endpoint and parse the metrics to retrieve them into a nice usable open metrics format
- [commit 3](https://github.com/cilium/cilium-cli/pull/1575/commits/86526acefae7242d946281b0719f1fa494cb3fc5) `policy.go` is a pretty long file now, let's move `Result` with its associated methods in its own file.
- [commit 4](https://github.com/cilium/cilium-cli/pull/1575/commits/36757130512333e191d4dd380e00a3335eff5f94) Make the call to collect metrics in the action
- [commit 5](https://github.com/cilium/cilium-cli/pull/1575/commits/7d5a902f35d8cded916db57600f3ffee6d76dae8) Validate the metrics by defining a new type `metricsCompareFunc` which will compare metrics in a generic way.
- [commit 6](https://github.com/cilium/cilium-cli/pull/1575/commits/18bb78f42845fcedddbf9af6a12d6e562c1f44c4) Add a new scenario very basic with a simple metric to verify the feature is working

## Usage

Implement the operation you want to check on your metrics on `Result`:
```go
func (r Result) ExpectMetricsIncrease(source MetricsSource, metrics ...string) Result {
```

Then use it directly in your expectations by passing a MetricSource type (cilium-agent, cilium-operator...) and the name of the metric you want to check:
```go
return check.ResultOK.ExpectMetricsIncrease(ct.CiliumAgentMetrics(), "cilium_forward_count_total")
```

### CODE REVIEW

File: connectivity/check/policy.go [hunk 1]
```diff
@@ -221,6 +222,31 @@ func authRequiredDropReason(flow *flowpb.Flow) bool {
...
 
+// expectMetricsToIncrease compares metrics retrieved before any action were run and after;
+// may return an error if metrics did not increase.
+func expectMetricsToIncrease(metrics []string) metricsCompareFunc {
+	return func(before, after map[string]*dto.MetricFamily) error {
+		for _, metricName := range metrics {
+			bValue, ok := before[metricName]
+			if !ok {
+				// TODO DONIA: should I return an error here, should I store the error and range over all the metrics?
+				return fmt.Errorf("failed to retrieve metric %s in first retrieved metrics", metricName)
```
Comment (gandro) [inline connectivity/check/policy.go:233]: Yeah, I think it's better to use `errors.Join` here to report on all metrics. It will make debugging test runs easier if you have more information on what worked and what didn't.

File: connectivity/check/policy.go [hunk 2]
```diff
@@ -333,6 +359,12 @@ var (
 		Drop:     true,
 		ExitCode: ExitCurlHTTPError,
 	}
+
+	// ResultOKWithMetricForwardCountTotal expects OK and
+	// a metric of the number of total of forward count increases.
+	ResultOKWithMetricForwardCountTotal = Result{
```
Comment (gandro) [inline connectivity/check/policy.go:365]: I think we should make this a method on `Result`, that also takes in the metric names as an argument. So in expectations, you could use something like the following in `suite.go`:

```go
WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
	return check.ResultOK.ExpectMetricsIncrease("cilium_forward_count_total", "cilium_foo_bar_total"), check.ResultOk
})
```

This allows you to easily test for drops too:
```go
WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
	return check.ResultDrop.ExpectMetricsIncrease("cilium_dropped_count_total"), check.ResultDrop
})
```

File: connectivity/check/policy.go [hunk 3]
```diff
@@ -399,6 +437,9 @@ func (r Result) String() string {
 		ret += "-"
 		ret += r.HTTP.Status
 	}
+	if r.MetricsCompareFunc != nil {
+		ret += fmt.Sprintf("-metricsCompareFunc")
```
Comment (gandro) [inline connectivity/check/policy.go:441]: nit:

```suggestion
		ret += "-MetricsCompareFunc"
```

File: k8s/dialer.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		return nil, err
+	}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, req.URL())
+	stopChan, readyChan := make(chan struct{}, 1), make(chan struct{}, 1)
+	if len(p.Addresses) == 0 {
+		p.Addresses = []string{"localhost"}
+	}
+
+	pw, err := portforward.NewOnAddresses(dialer, p.Addresses, p.Ports, stopChan, readyChan, os.Stdout, os.Stderr)
```
Comment (gandro) [inline k8s/dialer.go:54]: I think the output of Stdout and Stderr are only useful when debugging. To avoid this pouting the output, we should make those two `io.Writers` a parameter in `PortForwardParameters`.

And then in `collectPrometheusMetrics`, implement the `io.Writer` interface on top of the logging infrastructure in `connectivity/check/logging.go`. For example:

```
type debugWriter struct {
   ct *ConnectivityTest
}

func (d *debugWriter) Write(b []byte) (int, error) {
   d.ct.Debug(string(b))
   return len(b), nil
}

type warnWriter struct {
   ct *ConnectivityTest
}

func (w *warnWriter) Write(b []byte) (int, error) {
   w.ct.Warn(string(b))
   return len(b), nil
}
```

And then pass in a debugWriter instead of stdout, and a warnWriter instead of stderr.

File: connectivity/check/check.go [hunk 5]
```diff
@@ -72,6 +72,9 @@ type Parameters struct {
 
 	CollectSysdumpOnFailure bool
 	SysdumpOptions          sysdump.Options
+
+	// PrometheusServeAddr is the exposed port on cilium agent to retrieve metrics.
+	PrometheusServeAddr string // TODO DONIA: Not sure where to store it
```
Comment (gandro) [inline connectivity/check/check.go:77]: I think should define an interface here, that tells us what pods to collect from (e.g. agent, operator, hubble etc) on which port.

It could be something like:

```suggestion
	CollectMetrics []MetricsSource
```

Where the interface is:

```go
type MetricsSource interface {
   func Pods() ([]*corev1.Pod, error)
   func Port() string
}
```

And then implement this interface e.g. for the agent metrics. The expectations also should specify where they want to collect metrics from and what metrics they actually care about.

```go
WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
	return check.ResultOK.ExpectMetricsIncrease(a.CiliumAgentMetrics(), "cilium_forward_count_total"), check.ResultOk
})
```

Where as `CiliumAgentMetrics` defines a `MetricsSource`. This is just a rough sketch and needs to be fleshed out (i.e. I don't expect my proposal to work as is), but generally speaking the idea here should be that expectations declare what metrics they are interested in, and that in turn tells the framework not only that it needs to collect metrics, but also where it needs to collect metrics from.

Happy to discuss this further offline.

**Review by gandro — CHANGES_REQUESTED** [review]
Awesome work, thanks a lot! I've left some feedback regarding the open TODOs.

File: connectivity/check/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParsePromMetrics(t *testing.T) {
+	input := `
+# HELP isovalent_external_dns_proxy_policy_l7_total Number of total proxy requests handled
+# TYPE isovalent_external_dns_proxy_policy_l7_total counter
+isovalent_external_dns_proxy_policy_l7_total{rule="forwarded"} 132
+isovalent_external_dns_proxy_policy_l7_total{rule="received"} 132
+`
```
Comment (tklauser) [inline connectivity/check/metrics_test.go:18]: Could we please use metrics here that are not proprietary, i.e. that are available in an  Cilium OSS version?

Comment (doniacld) [reply connectivity/check/metrics_test.go:18]: Oups! Yes of course!

File: connectivity/check/policy.go (hunk 1 above)
Comment (doniacld) [reply connectivity/check/policy.go:233]: Fixed, thanks!

File: connectivity/check/metrics_test.go (hunk 6 above)
Comment (doniacld) [reply connectivity/check/metrics_test.go]: Fixed.

File: connectivity/check/policy.go (hunk 3 above)
Comment (doniacld) [reply connectivity/check/policy.go:444]: Fixed.

File: k8s/dialer.go (hunk 4 above)
Comment (doniacld) [reply k8s/dialer.go:54]: Fixed but I still have a log from the k8s go client port forward library:
https://github.com/kubernetes/client-go/blob/v0.27.1/tools/portforward/portforward.go#L312

The line of code:
```
runtime.HandleError(fmt.Errorf("error accepting connection on port %d: %v", port.Local, err))
```

The log (and there some lines...):
```
E0508 15:06:18.542005   72309 portforward.go:417] error closing listener: close tcp4 127.0.0.1:64508: use of closed network connection
```

I would be happy to discuss how to handle it.

File: connectivity/check/policy.go (hunk 2 above)
Comment (doniacld) [reply connectivity/check/policy.go]: I did some refactoring to isolate more Result and its associated methods, which might not be the best idea. It is in a new commit (the last one) for the moment but I will rework the history if we go with that.

File: k8s/dialer.go (hunk 4 above)
Comment (doniacld) [reply k8s/dialer.go]: Fixed.

File: connectivity/check/policy.go (hunk 2 above)
Comment (doniacld) [reply connectivity/check/policy.go]: Fixed.

File: k8s/dialer.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	ErrOut io.Writer
+}
+
+// PortForwardResult are the ports that have been forwarded.
+type PortForwardResult struct {
+	ForwardedPorts []ForwardedPort
+}
+
+// PortForward executes in a goroutine a port forward command.
+// To kill the thread, you should use the context by cancelling it.
```
Comment (gandro) [inline k8s/dialer.go:46]: nit:

`To stop the port-forwarding, use the context by cancelling it`

File: connectivity/check/metrics.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+type promMetricsPerSource map[string]promMetricsPerNode
+
+// promMetricsPerNode stores for each node Prometheus Metrics.
+type promMetricsPerNode map[string]promMetricsFamily
+
+// promMetricsFamily holds Prometheus metrics per metric name.
+type promMetricsFamily map[string]*dto.MetricFamily
+
+// collectPrometheusMetrics retrieves the Prometheus metrics by
+// port-forwarding the Prometheus port on a Cilium node and calling /metrics endpoint.
```
Comment (gandro) [inline connectivity/check/metrics.go:30]:
```suggestion
// port-forwarding the Prometheus port of each source pod and calling /metrics endpoint.
```

File: connectivity/check/metrics.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		return nil, fmt.Errorf("failed to parse metrics: %w", err)
+	}
+
+	return metrics, nil
+}
+
+// parseMetrics transforms the response from the call to prometheus metric endpoint
+// into a dto model MetricFamily.
+func parseMetrics(content []byte) (promMetricsFamily, error) {
+	reader := bytes.NewReader(content)
```
Comment (gandro) [inline connectivity/check/metrics.go:94]: Given that `resp.Body` is already a reader, we should probably pass in `resp.Body` directly here, instead of turning it into `content []byte` and then back into a reader.

File: k8s/dialer.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
+package k8s
```
Comment (gandro) [inline k8s/dialer.go:1]: Missing copyright header:

```
// SPDX-License-Identifier: Apache-2.0
// Copyright Authors of Cilium
```

File: connectivity/check/metrics.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
+package check
```
Comment (gandro) [inline connectivity/check/metrics.go:1]: Missing copyright header:

```
// SPDX-License-Identifier: Apache-2.0
// Copyright Authors of Cilium
```

File: connectivity/check/metrics_test.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
+package check
```
Comment (gandro) [inline connectivity/check/metrics_test.go:1]: Missing copyright header

File: connectivity/check/metricssource.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
+package check
```
Comment (gandro) [inline connectivity/check/metricssource.go:1]: Missing copyright header

File: connectivity/check/metricssource_test.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
+package check
```
Comment (gandro) [inline connectivity/check/metricssource_test.go:1]: Missing copyright header

File: connectivity/check/metricssource.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+	// Retrieve the container port value for Prometheus.
+	for _, p := range ciliumPods {
+		// parse all the containers
+		for _, c := range p.Pod.Spec.Containers {
+			if c.Name == components.CiliumAgentName {
+				// parse all the container ports
+				for _, port := range c.Ports {
+					if port.Name == prometheusContainerPortName {
+						source.Port = fmt.Sprintf("%d", port.ContainerPort)
```
Comment (gandro) [inline connectivity/check/metricssource.go:43]: consider exiting the `range ciliumPods` loop early here once we found a port - alternatively, it might be cleaner to actually `(pod, port)` pairs in the `MetricsSouce`.

File: connectivity/check/metricssource.go [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+					}
+				}
+			}
+		}
+	}
+
+	// Prometheus port was not find, let's return an empty MetricsSource.
+	if source.Port == "" {
+		return MetricsSource{}
```
Comment (gandro) [inline connectivity/check/metricssource.go:52]: Consider defining a `EmptyMetricsSource` constant for this, so callers can compare to the constant to determine if the metrics source exists or not.

File: connectivity/check/action.go [hunk 17]
```diff
@@ -139,6 +143,24 @@ func (a *Action) Run(f func(*Action)) {
 	// Emit unbuffered progress indicator.
 	a.test.progress()
 
+	// Retrieve Prometheus metrics only if there are expectations.
+	if a.expIngress.Metrics != nil {
```
Comment (gandro) [inline connectivity/check/action.go:147]: No need for a nil check if we only iterate in the loop body

File: connectivity/check/policy.go [hunk 18]
```diff
@@ -24,43 +24,12 @@ import (
 	"github.com/cilium/cilium-cli/k8s"
 )
 
-type ExitCode int16
```
Comment (gandro) [inline connectivity/check/policy.go:27]: Consider moving this refactor in a separate commit to make it easier on reviewers

File: connectivity/check/result.go [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+	if r.HTTP.URL != "" {
+		ret += "-"
+		ret += r.HTTP.URL
+	}
+	if r.HTTP.Status != "" {
+		ret += "-"
+		ret += r.HTTP.Status
+	}
+	if r.Metrics != nil {
+		ret += "-metricsCompareFunc"
```
Comment (gandro) [inline connectivity/check/result.go:122]: I think we should print the metric source names here.

File: connectivity/check/action.go [hunk 20]
```diff
@@ -942,3 +977,68 @@ func (a *Action) validateFlowsForPeer(ctx context.Context, reqs []filters.FlowSe
...
+	ctx, cancel := context.WithTimeout(ctx, metricsCollectionTimeout)
+	defer cancel()
+
+	// ticker here is used as the interval duration between before the next retry.
+	ticker := time.NewTicker(metricCollectionIntervalRetry)
+	defer ticker.Stop()
+
+	for {
+		// Collect the new metrics.
+		newMetrics, err := a.collectPrometheusMetrics(result.Source)
```
Comment (gandro) [inline connectivity/check/action.go:1018]: As an optimization, we could consider adding a `collectPrometheusMetricsForNode` function, that only port-forwards to a particular node

File: connectivity/tests/pod.go [hunk 21]
```diff
@@ -59,6 +59,9 @@ func (s *podToPod) Run(ctx context.Context, t *check.Test) {
 
 					a.ValidateFlows(ctx, client, a.GetEgressRequirements(check.FlowParameters{}))
 					a.ValidateFlows(ctx, echo, a.GetIngressRequirements(check.FlowParameters{}))
+
+					a.ValidateMetrics(ctx, echo, a.GetIngressMetricsRequirements())
+					a.ValidateMetrics(ctx, echo, a.GetEgressMetricsRequirements())
```
Comment (gandro) [inline connectivity/tests/pod.go:64]: We probably want to add this to all tests where it makes sense (check for calls to `a.ValidateFlows` on where to add)

File: connectivity/check/metrics.go [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+	// Call the k8s dialer to port forward,
+	// a random port will be generated to avoid conflict.
+	res, err := pod.K8sClient.PortForward(ctx, p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to port forward: %w", err)
+	}
+
+	// Call the metrics path on the retrieved local port.
+	url := fmt.Sprintf(metricsURLFormat, res.ForwardedPorts[0].Local)
+	resp, err := http.Get(url)
```
Comment (gandro) [inline connectivity/check/metrics.go:71]: Here is a linting error on this:
```
G107: Potential HTTP request made with variable url (gosec)
```

I think our usage is safe, so consider adding a ` //nolint:gosec` annotation here

**Review by gandro — CHANGES_REQUESTED** [review]
Awesome! High-level this looks good to me. I did leave some minor feedback that should be addressed before we can merge. Thanks for working on this!

File: connectivity/check/metrics.go (hunk 22 above)
Comment (doniacld) [reply connectivity/check/metrics.go:71]: Fixed.

File: connectivity/check/metrics.go (hunk 9 above)
Comment (doniacld) [reply connectivity/check/metrics.go:94]: Good catch! 👁️ 👓

File: connectivity/check/metricssource.go (hunk 15 above)
Comment (doniacld) [reply connectivity/check/metricssource.go:43]: I added the early exit, but I will be happy to discuss the interest of pairing pod, port :)

File: connectivity/check/metricssource.go (hunk 16 above)
Comment (doniacld) [reply connectivity/check/metricssource.go:52]: Cannot define a constant on a non-primitive type, should I create a function?

File: connectivity/tests/pod.go (hunk 21 above)
Comment (doniacld) [reply connectivity/tests/pod.go:64]: I am pretty unsure of what I've done here, so please be careful in the review :)

File: connectivity/check/action.go (hunk 20 above)
Comment (doniacld) [reply connectivity/check/action.go:1015]: Fixed.

File: connectivity/check/metricssource.go (hunk 16 above)
Comment (doniacld) [reply connectivity/check/metricssource.go:51]: Done! Using a function on MetricsSource.

File: connectivity/check/metricssource.go (hunk 15 above)
Comment (doniacld) [reply connectivity/check/metricssource.go:41]: For the record, we decided with @gandro to not pair a pod to a port for the moment but we could have a usecase in the future (e.g. cluster mesh use-case: different configuration of ports for a component through different clusters)

File: connectivity/check/metrics.go [hunk 23]
```diff
@@ -53,7 +53,7 @@ func (a *Action) collectPrometheusMetrics(source MetricsSource) (promMetricsPerN
 }
 
 // collectMetricsForPod retrieves the metrics for one pod.
-func (a *Action) collectMetricsForPod(pod Pod, port string) (map[string]*dto.MetricFamily, error) {
```
Comment (gandro) [inline connectivity/check/metrics.go]: nit: This change here seems to be part of the wrong commit

File: connectivity/check/result.go [hunk 24]
```diff
@@ -105,8 +124,24 @@ func (r Result) String() string {
...
+	}
 	if r.ExitCode >= 0 && r.ExitCode <= 255 {
 		ret += fmt.Sprintf("-exit(%d)", r.ExitCode)
 	}
 	return ret
 }
+
+func displayMetrics(metrics []MetricsResult) string {
+	var ret string
+	ret += "-Sources("
```
Comment (gandro) [inline connectivity/check/result.go:139]: nit: Let's call this `MetricSources(` or `Metrics(`. It's not clear from context otherwise that this refers to metrics.

**Review by gandro — COMMENTED** [review]
Awesome, overall looks good to me! A couple minor nits remaining

File: connectivity/check/metrics.go (hunk 23 above)
Comment (doniacld) [reply connectivity/check/metrics.go]: Oupsy! Fixed!

File: connectivity/check/action.go [hunk 25]
```diff
@@ -1027,7 +1027,7 @@ func (a *Action) validateMetric(ctx context.Context, node string, result Metrics
 	ctx, cancel := context.WithTimeout(ctx, metricsCollectionTimeout)
 	defer cancel()
 
-	// ticker here is used as the interval duration between before the next retry.
```
Comment (gandro) [inline connectivity/check/action.go]: nit: I think typo fix should be in an earlier commit

**Review by gandro — APPROVED** [review]
Looks good to me! Two minor (non-blocking) nits remaining

File: connectivity/check/metrics.go [hunk 26]
```diff
@@ -115,3 +115,23 @@ func parseMetrics(reader io.Reader) (promMetricsFamily, error) {
...
+func metricsIncrease(mf1, mf2 dto.MetricFamily) error {
+	metrics1 := mf1.GetMetric()
+	metrics2 := mf2.GetMetric()
+
+	if len(metrics1) != len(metrics2) {
+		return fmt.Errorf("metric %s has different length metrics 1: %d and metrics 2: %d", mf1.GetName(), len(metrics1), len(metrics2))
+	}
+
+	for i := range metrics1 {
+		value1 := metrics1[i].GetCounter().GetValue() // Here we assume that metrics are of Counter type.
```
[the remark this answers is not in the document]
Comment (doniacld) [reply connectivity/check/metrics.go:124]: Is it fine to add these checks?

```go
if metrics1[i].GetCounter() == nil {
	return fmt.Errorf("metric %s is not a Counter: %v", mf1.GetName(), metrics1[i])
}
if metrics2[i].GetCounter() == nil {
	return fmt.Errorf("metric %s is not a Counter: %v", mf1.GetName(), metrics2[i])
}
```

**Review by brlbil — APPROVED** [review]
The CI part is 👍 Awesome work, thanks a lot!

**Review by youngnick — APPROVED** [review]

**Review by tklauser — APPROVED** [review]
Nice work!
